### PR TITLE
Create stat name handler not supported rule

### LINKED ISSUE #11045: Create StatNameHandlerNotSupportedRule to ease upgrade to Airflow 2.0
This issue is part of #8765

## Rule
Create `StatNameHandlerNotSupportedRule` which corresponds to
> Drop plugin support for stat_name_handler

entry in UPDATING.md. This rule should allow users to check if their current configuration needs any adjusting
before migration to Airflow 2.0.

## How to guide
To implement a new rule, create a class that inherits from `airflow.upgrade.rules.base_rule.BaseRule`.
It will be auto-registered and used by `airflow upgrade-check` command. The custom rule class has to have `title`,
`description` properties and should implement `check` method which returns a list of error messages in case of
incompatibility.

For example:
https://github.com/apache/airflow/blob/ea36166961ca35fc51ddc262ec245590c3e236fb/airflow/upgrade/rules/conn_type_is_not_nullable.py#L25-L42

**Remember** to open the PR against `v1-10-test` branch.

### PR DESCRIPTION
**Pull request by FHoffmannCode**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/upgrade/rules/stat_name_handler_not_supported.py
Comment (mik-laj) [inline airflow/upgrade/rules/stat_name_handler_not_supported.py]: It is okay that there is this attribute. This allows us to create plugins that are backwards compatible.

We should check if the `stat_name_handler` option has been set. If so, skip role. If not, check to see if there are any plugins that use it. If not, skip any further processing. If so, return a message to the user.

Comment (FHoffmannCode) [reply airflow/upgrade/rules/stat_name_handler_not_supported.py]: ok, fixd

**Review by turbaszek — APPROVED** [review]
LGTM

**Comment by turbaszek** [comment]
@FHoffmannCode can you please rebase as it seems that there were changes in the meantime?

**Comment by potiuk** [comment]
@FHoffmannCode Can you please rebase to latest v1-10-test  ? It should work now.

**Comment by FHoffmannCode** [comment]
@potiuk @turbaszek rebase done

File: airflow/upgrade/rules/stat_name_handler_not_supported.py
Comment (turbaszek) [inline airflow/upgrade/rules/stat_name_handler_not_supported.py]: Please remove the indentation:
```
stat_name_handler field in AirflowTestPlugin is not supported.
--------------------------------------------------------------
    stat_name_handler field is no longer supported in AirflowTestPlugin.
    Instead there is stat_name_handler option in [scheduler] section of airflow.cfg.
    This change is intended to simplify the statsd configuration.
```

File: tests/upgrade/rules/test_stat_name_handler_not_supported.py
Comment (turbaszek) [inline tests/upgrade/rules/test_stat_name_handler_not_supported.py]: It will be simpler to use `conf_vars`tests.tests_utils.config.contextmanger`

File: tests/test_utils/mock_plugins.py
Comment (turbaszek) [inline tests/test_utils/mock_plugins.py]: Wouldn't it be easier to just mock `airflow.upgrade.rules.stat_name_handler_not_supported.plugins_manager` and then do 
```
mock_plugin_manager.plugins = [MagicMock(stat_name_handler=True)]
```
WDYT?

File: tests/upgrade/rules/test_stat_name_handler_not_supported.py
Comment (FHoffmannCode) [reply tests/upgrade/rules/test_stat_name_handler_not_supported.py]: indeed

Comment (FHoffmannCode) [reply tests/upgrade/rules/test_stat_name_handler_not_supported.py]: fixd

File: airflow/upgrade/rules/stat_name_handler_not_supported.py
Comment (FHoffmannCode) [reply airflow/upgrade/rules/stat_name_handler_not_supported.py]: fixd

File: tests/test_utils/mock_plugins.py
Comment (FHoffmannCode) [reply tests/test_utils/mock_plugins.py]: indeed its easier - fixd

**Comment by turbaszek** [comment]
@FHoffmannCode there's still to be unrelated commits in this PR :<

**Comment by FHoffmannCode** [comment]
@turbaszek oops my bad, its fixed

**Comment by turbaszek** [comment]
@FHoffmannCode can you please rebase?

**Comment by FHoffmannCode** [comment]
@turbaszek sure, done

**Comment by mik-laj** [comment]
> 53 successful, 40 failing, 2 cancelled, and 5 skipped checks
CI is sad. Can you fix it?

**Comment by FHoffmannCode** [comment]
@mik-laj tests that are failing are not related to my PR, pipeline does not pass on `v1-10-test` branch.

**Comment by mik-laj** [comment]
@FHoffmannCode We need to fix the v1-10-test branch before we can merge any change.

**Comment by turbaszek** [comment]
@FHoffmannCode would you mind rebasing and changing the target base due to https://github.com/apache/airflow/pull/11719#issuecomment-714258732

**Comment by FHoffmannCode** [comment]
@turbaszek @mik-laj 

I've just cherry-picked my commit to `v1-10-stable` pushed the code and changed base branch to `apache/v1-10-stable`

**Review by mik-laj — APPROVED** [review]

File: airflow/upgrade/rules/stat_name_handler_not_supported.py
Comment (kenluck2001) [inline airflow/upgrade/rules/stat_name_handler_not_supported.py:39]: The for loop is not using plugin.The any function is not clear in its use.

Comment (turbaszek) [reply airflow/upgrade/rules/stat_name_handler_not_supported.py:39]: @kenluck2001 we check if any `plugin` in `plugin_manager` has `stat_name_handler` which is correct behaviour. Or do I miss something?

**Comment by turbaszek** [comment]
Rebased and I will merge once the tests pass

**Comment by kaxil** [comment]
Can you please rebase your PR on latest v1-10-stable please, the tests were running previously -- rebasing should fix it

**Comment by RosterIn** [comment]
@FHoffmannCode test is failing
```
=================================== FAILURES ===================================
___________ TestAirflowMacroPluginRemovedRule.test_bad_file_failure ____________

self = <tests.upgrade.rules.test_airflow_macro_plugin_removed.TestAirflowMacroPluginRemovedRule testMethod=test_bad_file_failure>
mock_list_files = <MagicMock name='list_py_file_paths' id='140705210084688'>

    def test_bad_file_failure(self, mock_list_files):
        # Write a binary file
        with NamedTemporaryFile("wb+", suffix=".py") as temp_file:
            mock_list_files.return_value = [temp_file.name]
            temp_file.write(b"{\x03\xff\x00d")
            temp_file.flush()
    
            rule = AirflowMacroPluginRemovedRule()
            msgs = rule.check()
>           assert 1 == len(msgs)
E           AssertionError: assert 1 == 0
E            +  where 0 = len([])

tests/upgrade/rules/test_airflow_macro_plugin_removed.py:92: AssertionError
```

**Pull request closed without merging**
